### PR TITLE
PAB v2

### PR DESCRIPTION
**Pull request by j-mueller**
This turned into a complete makeover for the PAB.

SCP-1897, SCP-1959, SCP-1815

### Changes

* Add a new, STM-based implementation of `ProcessContractOutboxes` to `Plutus.PAB.Core.ContractInstance`
* Add STM-based request handlers for waiting effects such as `AwaitSlot` and `NextTxAt`
* Isolate the `eventful` stuff from most of the code, moving all database related funtions to the `Plutus.PAB.Db.Eventful` module
* Add a new simulator in `Plutus.PAB.Simulator` that actually runs the STM-based contracts in `IO`
  * There is no need to have a PAB-specific `MultiAgent` effect, we can just run everything in `IO` and the contract instances get their own threads
  * With the new simulator you can now really wait for things like endpoints, accumulated state of contracts, etc. using `STM` while the rest of the system keeps on ticking in the background. See for example [this line](https://github.com/input-output-hk/plutus/pull/2801/files#diff-72871f6bd287429c144b740a1aa741759a39ba30715b680f8c15dec943a8dc8aR114) where we wait until the currency has been produced on the chain
  * The new simulator prints sensible debug output to stdout
  * The test suite has been adapted to use the new simulator
* Flesh out the new REST API
  * You can now get updates for each instance via a websocket
  * The API can be served from within the simulator, so you can program it to do some setup first (see [here](https://github.com/input-output-hk/plutus/pull/2801/files#diff-3d6a3d3c996a1fce0ee1c60be3eef376ec5d774100fac42f6d84e6a653e05130R187-R190))
    * I imagine this could be useful for testing / repl based development
* Deleted a couple of unused types; the PAB code was quite messy :(

### TODO

* Generalise the simulator to work with any type that's an instance of `PABContract` - right now it only does `TestContracts`, but it shouldn't be hard to change that (maybe next PR)
  * Then we could use it for the two marlowe contracts
* Update PSGEnerator and Purescript frontend
  * Probably need to delete a lot of stuff although I tried to keep the essential bits
  * We need to revisit the PAB UI anyway.

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by j-mueller** [comment]
Fixes SCP-1959

**Review by raduom — APPROVED** [review]
My only observations would be that deleting things that only seem to work is probably a very good idea. They simplify the code, which is going to help a lot with adoption, and there is less to review.

In that spirit, I am not sure why we are keeping the old polling interface. I suspect it is still used in the `PAB` code and will eventually be removed as we move the `PAB` code to STM.

The rest of the code looks good to me, but I did not spend enough time to gain an in-depth understanding of what is going here, so I would suggest, for now, to wait for a second opinion.

**Comment by raduom** [comment]
I would also like to know if this implementation is faster than the old one. It's probably a lot snappier (no timeouts), but I would expect that the TPS will converge to a fairly similar value for polling vs STM (because we already batched events in the old implementation, I think). If that is not the case I would like to understand why it is faster, if you have the time.

File: plutus-pab/src/Plutus/PAB/Core/ContractInstance/Polling.hs [hunk 1]
```diff
@@ -0,0 +1,365 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
```
[the remark this answers is not in the document]
Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Core/ContractInstance/Polling.hs:1]: Yes. Maybe I should just delete the polling interface, as you said.

**Comment by j-mueller** [comment]
> I would also like to know if this implementation is faster than the old one

I'm interested in that too. But I suspect it will only matter if we have a number of active contract instances. Because then most of them will be sleeping, whereas previously we would go through all the instances in every slot and see if they were waiting for anything.

> deleting things that only seem to work is probably a very good idea

Yeah I think I'm just going to do that now. It'll make this PR bigger, but the result should be a lot simpler.

**Comment by j-mueller** [comment]
Seems like all I do these days is writing simulators. But this one actually feels quite nice to use.

File: plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs [hunk 2]
```diff
@@ -11,102 +11,24 @@
...
-        , _chainIndexState = CI.initialAppState
-        , _signingProcessState = defaultSigningProcess wallet
-        , _agentEventState = emptyEventMap
-        }
-
-agentState :: Wallet.Wallet -> Lens' (Map Wallet AgentState) AgentState
-agentState wallet = at wallet . anon (emptyAgentState wallet) (const False)
+import           Wallet.Emulator.MultiAgent               (EmulatorEvent)
 
+-- FIXME: Replace with PABLogMsg?
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs:31]: Is this supposed to remain here?

File: plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs [hunk 3]
```diff
@@ -115,6 +37,8 @@ data PABMultiAgentMsg =
     | ContractInstanceLog (ContractInstanceMsg TestContracts)
     | CoreLog (CoreMsg TestContracts)
     | RuntimeLog ContractRuntimeMsg
+    | UserLog T.Text
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs:40]: Yeah. I was thinking of doing something like this, also. :)

File: plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs [hunk 4]
```diff
@@ -11,102 +11,24 @@
 {-# LANGUAGE TypeApplications  #-}
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Effects/MultiAgent.hs:11]: Where do we need this?

File: plutus-pab/src/Plutus/PAB/Types.hs [hunk 5]
```diff
@@ -60,9 +45,9 @@ data PABError
     | MetadataError Metadata.MetadataError
     | ChainIndexError ClientError
     | WalletError WalletAPIError
-    | ContractCommandError Int Text
+    | ContractCommandError Int Text -- ?
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Types.hs:48]: You should either remove this or expand on it. :)

File: plutus-pab/src/Plutus/PAB/Webserver/Handler.hs [hunk 6]
```diff
@@ -202,34 +181,38 @@ parseContractId t =
...
-       , Member (LogMsg LM.ContractExeLogMsg) effs
        , Member (Error PABError) effs
        , Member (LogMsg UnStringifyJSONLog) effs
        , Member (LogMsg (Instance.ContractInstanceMsg ContractExe)) effs
+       , Instance.AppBackendConstraints ContractExe m appBackend
+       , LastMember m effs
+       , Member (Reader InstancesState) effs
+       , Member (ContractDefinitionStore ContractExe) effs
+       , Member (ContractStore ContractExe) effs
+       , LastMember m (Reader ContractInstanceId ': appBackend)
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Webserver/Handler.hs:196]: Shouldn't this constraint be the same as `LastMember m appBackend`. Just curious as to why you need to prepend `ContractsInstanceId`.

File: plutus-pab/src/Plutus/PAB/Webserver/Server.hs [hunk 7]
```diff
@@ -14,78 +15,96 @@ module Plutus.PAB.Webserver.Server
...
-          -> Eff (LogMsg LM.ContractExeLogMsg
-                 ': LogMsg (ContractInstanceMsg ContractExe)
-                 ': LogMsg WebSocketLogMsg
-                 ': LogMsg UnStringifyJSONLog
-                 ': AppBackend _) a
-          -> Handler a
-asHandler trace logConfig config =
-    Handler . ExceptT . fmap (first decodeErr)
-      . runApp trace logConfig config
+runHandler :: InstancesState -> Trace IO LM.PABLogMsg -> CM.Configuration -> Config -> Eff (LogMsg LM.ContractExeLogMsg ': LogMsg (ContractInstanceMsg ContractExe) ': LogMsg WebSocketLogMsg ': LogMsg UnStringifyJSONLog ': AppBackend _) a -> IO (Either PABError a)
```
Comment (raduom) [inline plutus-pab/src/Plutus/PAB/Webserver/Server.hs:59]: No sure if this is an issue, but this line is really long.

File: plutus-pab/test/Plutus/PAB/Events/ContractSpec.hs [hunk 8]
```diff
@@ -41,7 +43,7 @@ jsonTests =
                       { aeDescription = EndpointDescription (symbolVal (Proxy @"guess"))
                       , aeMetadata    = Nothing
                       }
-              response :: Either String ContractHandlersResponse
+              response :: Either String ContractHandlerRequest
```
Comment (raduom) [inline plutus-pab/test/Plutus/PAB/Events/ContractSpec.hs:46]: Is it ok for the `response` to have type `...Request`?

**Review by raduom — APPROVED** [review]

**Comment by gilligan** [comment]
At `+3,653 −2,866 ` it's frankly impossible to me to review this adequately ;/

File: plutus-pab/src/Plutus/PAB/Webserver/WebSocket.hs [hunk 9]
```diff
@@ -2,175 +2,114 @@
...
-    -> Eff effs a
-    -> (a -> Eff effs ())
-    -> Eff effs ()
-watchForChanges time query notify = go Nothing
+contractInstanceUpdates :: forall t env. ContractInstanceId -> PendingConnection -> PABAction t env ()
+contractInstanceUpdates contractInstanceId pending = do
+    Core.PABRunner{Core.runPABAction} <- Core.pabRunner
+    liftIO $ do
+        connection <- WS.acceptRequest pending
+        handle disconnect . WS.withPingThread connection 30 (pure ()) $ fmap (either (error . show) id) . runPABAction $ sendUpdatesToClient contractInstanceId connection
```
Comment (gilligan) [inline plutus-pab/src/Plutus/PAB/Webserver/WebSocket.hs:78]:
```
handle disconnect . WS.withPingThread connection 30 (pure ()) $ fmap (either (error . show) id) . runPABAction $ sendUpdatesToClient contractInstanceId connection
```

o____O

what?

:sweat_smile:

File: plutus-pab/src/Plutus/PAB/Effects/Contract/ContractExe.hs [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+    | ProcessExitFailure String
+    | ContractResponse String
+    | Migrating
+    | InvokingEndpoint String Value
+    | EndpointInvocationResponse [Text]
+    | ContractExePABError PABError
+    deriving stock (Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+instance Pretty ContractExeLogMsg where
```
Comment (gilligan) [inline plutus-pab/src/Plutus/PAB/Effects/Contract/ContractExe.hs:125]: Seeing all the new/changed msg data types and respective instances for logging added i'd like to mention something that I pointed out in some chat before: I don't think we'll ever get consistent logging with all of those. Too many different instances in too many different places for that.

I'm also not convinced it is useful or helpful to be able to track what is being logged where in the effect system (while this on the other hand leads to rather tight coupling of types).

This PR is far too huge already but I think there should be _one_  logging data type with some mandatory fields (like say `component` and `isMock` and maybe some others i can't think of now) and some free-text field.

Comment (j-mueller) [reply plutus-pab/src/Plutus/PAB/Effects/Contract/ContractExe.hs:125]: Yeah I agree with you. We should keep the structured logging in the emulator (`EmulatorLogMsg`) but we don't really need it here in the PAB.

**Pull request merged**